File: app/src/main/java/com/example/thekhaeng/viewstatesavetest/view/java/CustomSwitch.java
package com.example.thekhaeng.viewstatesavetest.view.java;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.widget.Switch;

/**
 * Created by TheKhaeng
 */

public class CustomSwitch extends Switch {

    private int customState;

    public CustomSwitch( Context context ){
        super( context );
    }

    public CustomSwitch( Context context, AttributeSet attrs ){
        super( context, attrs );
    }

    public CustomSwitch( Context context, AttributeSet attrs, int defStyleAttr ){
        super( context, attrs, defStyleAttr );
    }

    public void setCustomState( int customState ){
        this.customState = customState;
    }

    @Override
    public Parcelable onSaveInstanceState(){
        Parcelable superState = super.onSaveInstanceState();
        SavedState ss = new SavedState( superState );
        ss.state = customState;
        return ss;
    }

    @Override
    public void onRestoreInstanceState( Parcelable state ){
        if (!(state instanceof SavedState)) {
            super.onRestoreInstanceState(state);
            return;
        }
        SavedState ss = (SavedState) state;
        super.onRestoreInstanceState(ss.getSuperState());
        setCustomState( ss.state );
    }

    private static class SavedState extends BaseSavedState{
        int state;

        SavedState( Parcelable superState ){
            super( superState );
        }

        private SavedState( Parcel in ){
            super( in );
            state = in.readInt();
        }

        @Override
        public void writeToParcel( Parcel out, int flags ){
            super.writeToParcel( out, flags );
            out.writeInt( state );
        }

        public static final Creator<SavedState> CREATOR
                = new Creator<SavedState>(){
            public SavedState createFromParcel( Parcel in ){
                return new SavedState( in );
            }

            public SavedState[] newArray( int size ){
                return new SavedState[size];
            }
        };
    }
}

File: app/src/main/java/com/example/thekhaeng/viewstatesavetest/view/kotlin/CustomSwitch.kt
package com.example.thekhaeng.viewstatesavetest.view.kotlin

import android.content.Context
import android.os.Build
import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.widget.Switch
import com.example.thekhaeng.viewstatesavetest.view.kotlin.base.ViewSavedState

/**
 * Created by TheKhaeng
 */

class CustomSwitch
    : Switch {

    private var customState: Int = 0

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
    }

    fun setCustomState(customState: Int) {
        this.customState = customState
    }

    override
    fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val ss = SavedState(superState)
        ss.state = customState
        return ss
    }


    override
    fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }
        super.onRestoreInstanceState(state.superState)
        setCustomState(state.state)
    }


    internal class SavedState : ViewSavedState {
        var state: Int = 0

        constructor(superState: Parcelable) : super(superState)

        constructor(`in`: Parcel) : super(`in`) {
            state = `in`.readInt()
        }

        override
        fun writeToParcel(out: Parcel, flags: Int) = with(out) {
            super.writeToParcel(out, flags)
            out.writeInt(state)
        }

        companion object {
            @JvmField
            val CREATOR: Parcelable.Creator<SavedState> = object : Parcelable.Creator<SavedState> {

                override
                fun createFromParcel(`in`: Parcel): SavedState? = SavedState(`in`)

                override
                fun newArray(size: Int): Array<SavedState?> = arrayOfNulls(size)
            }
        }
    }
}
